components/PullToRefresh.vue

```diff
@@ -0,0 +1,147 @@
+<docs>
+	# 下拉刷新
+
+	FIXME: 当前因为使用了TouchEvent只支持Chromium内核的浏览器，请考虑使用PointerEvent但是注意浏览器默认触摸事件的冲突问题。
+</docs>
+
+<script setup lang="ts">
+	const props = withDefaults(defineProps<{
+		/** 滚动容器 querySelector。 */
+		scrollContainerSelector?: string;
+		/** 滚动内容 querySelector。 */
+		scrollContentSelector?: string;
+	}>(), {
+		scrollContainerSelector: "#mainScroller",
+		scrollContentSelector: "#mainScroller > .content",
+	});
+	const refreshing = defineModel<boolean>();
+
+	const scrollContainer = ref<HTMLElement | null>(null);
+	const scrollContent = ref<HTMLElement | null>(null);
+	const indicatorWrapper = ref<HTMLElement>();
+
+	const REFRESH_THRESHOLD = 160;
+	const pulling = ref(false);
+	const startY = ref(0);
+	const pullDistance = ref(0);
+	const indicatorRotate = computed(() => {
+		if (refreshing.value) return 0;
+		if (pulling.value)
+			return Math.max(pullDistance.value / REFRESH_THRESHOLD, 0) * 360;
+		return 0;
+	});
+	const indicatorWrapperTranslateY = computed(() => {
+		if (refreshing.value) return 32;
+		if (pulling.value)
+			return map(pullDistance.value, 0, REFRESH_THRESHOLD, -48, 32);
+		return -48;
+	});
+
+	const onTouchMove = (e: TouchEvent) => {
+		if (scrollContainer.value?.scrollTop !== 0) return;
+		const currentY = e.touches[0].clientY;
+		if (currentY - startY.value > 0) {
+			pulling.value = true;
+			pullDistance.value = currentY - startY.value;
+			if (indicatorWrapper.value) indicatorWrapper.value.style.transition = "unset";
+			scrollContent.value!.style.translate = `0 ${pullDistance.value / 2}px`;
+			scrollContent.value!.style.transition = "unset";
+		}
+	};
+
+	const onTouchStart = (e: TouchEvent) => {
+		if (scrollContainer.value?.scrollTop !== 0) return;
+		startY.value = e.touches[0].clientY;
+		scrollContainer.value?.addEventListener("touchmove", onTouchMove as EventListener);
+	};
+
+	const onTouchEnd = () => {
+		if (pulling.value && pullDistance.value >= REFRESH_THRESHOLD) {
+			refreshing.value = true;
+			setTimeout(() => {
+				refreshing.value = false;
+			}, 1000); // 模拟刷新延迟
+		}
+		pulling.value = false;
+		pullDistance.value = 0;
+		if (indicatorWrapper.value) indicatorWrapper.value.style.transition = "";
+		scrollContent.value!.style.translate = "";
+		scrollContent.value!.style.transition = `translate 1s ${eases.easeOutSpringSnap}`;
+		scrollContainer.value?.removeEventListener("touchmove", onTouchMove as EventListener);
+	};
+
+	onMounted(() => {
+		scrollContainer.value = document.querySelector(props.scrollContainerSelector);
+		scrollContent.value = document.querySelector(props.scrollContentSelector);
+		if (scrollContainer) {
+			scrollContainer.value?.addEventListener("touchstart", onTouchStart as EventListener);
+			scrollContainer.value?.addEventListener("touchend", onTouchEnd as EventListener);
+		}
```

Comment:
你这段 `if (scrollContainer)` 判断了个寂寞啊，它恒为 true。

话说我封装了更强大的 useEventListener composable，为啥还要原始的 onMounted addEventListener + onBeforeUnmount removeEventListener？

```suggestion
		scrollContainer.value?.addEventListener("touchstart", onTouchStart as EventListener);
		scrollContainer.value?.addEventListener("touchend", onTouchEnd as EventListener);
```

Reply:
> 话说我封装了更强大的 useEventListener composable，为啥还要原始的 onMounted addEventListener + onBeforeUnmount removeEventListener？

因为需要在选择器执行完毕后再执行